sdk/core/core/src/az_http_policy.c

```diff
@@ -47,30 +56,102 @@ AZ_NODISCARD az_result az_http_pipeline_policy_uniquerequestid(
 
 AZ_NODISCARD az_result az_http_pipeline_policy_retry(
     az_http_policy * const p_policies,
+    void * const data,
     az_http_request_builder * const hrb,
     az_mut_span const * const response) {
+  (void)data;
   AZ_CONTRACT_ARG_NOT_NULL(p_policies);
   AZ_CONTRACT_ARG_NOT_NULL(hrb);
   AZ_CONTRACT_ARG_NOT_NULL(response);
   // Retry logic
   return az_http_pipeline_nextpolicy(p_policies, hrb, response);
 }
 
+enum { AZ_HTTP_POLICY_AUTH_BUFFER_SIZE = 2 * 1024 };
+
+// Being given
+// "https://NNNNNNNN.vault.azure.net/secrets/Password/XXXXXXXXXXXXXXXXXXXX?api-version=7.0", gives
+// back "https://vault.azure.net" (needed for authentication).
+static AZ_NODISCARD az_result
+az_auth_get_resource_url(az_span const request_url, az_span_builder * p_builder) {
+  az_url url = { 0 };
+  if (!az_succeeded(az_url_parse(request_url, &url))) {
+    return AZ_ERROR_ARG;
+  }
+
+  az_span domains[3] = { 0 };
+  size_t const ndomains = sizeof(domains) / sizeof(domains[0]);
+  for (size_t i = 0; i < ndomains; ++i) {
+    if (!az_succeeded(az_host_read_domain(&url.authority.host, &domains[i]))) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  // Add "https://"
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, url.scheme));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, AZ_STR("://")));
+
+  // We need to iterate in backwards order, because az_dns_read_domain() is going to give us domains
+  // in reverse order, i.e. "net, azure, vault" for "vault.azure.net"
+  for (size_t i = ndomains; i > 1;) { // This loop would add "vault.azure."
```

Comment:
For me, it's a kind of common pattern

- backward
```c
for (size_t i = size; i > 0;) {
  --i;
  ….array[i]...
}
``` 
- forward
```c
for (size_t i = 0; i < size; ++i) {
   … array[i]...
}
```

Reply:
For backward, you need to run `--i` before use. An alternative looks ugly and doesn't work for unsigned types.

```
for (int i = size - 1; i >= 0; --i) {
  ….array[i]...
}
```

Reply:
@vhvb1989 , Sergey is right. In my specific case it does not matter since my condition is `i > 1`, so my loop can be written in more regular way, but in general, for reverse loops, it is better to decrement `i` inside the loop:
1. I can be `unsigned` type, so if you have a loop for `(i = size; i >=0; --i)`, `i >= 0` condition is always true
2. When you have a loop `for (i = size - 1; i >= 5; --i)`, when `size` is `0`, `i` gets overflown, and you iterate from `uint.max` to `5`.